src/fides/api/ops/util/consent_util.py

```diff
@@ -34,9 +35,16 @@ def filter_privacy_preferences_for_propagation(
 
 def should_opt_in_to_service(
     system: Optional[System], privacy_request: PrivacyRequest
-) -> Optional[bool]:
+) -> Tuple[Optional[bool], List[PrivacyPreferenceHistory]]:
```

Comment:
definitely makes sense to couple these together as they are related and derived through the same operations, and i don't have any concrete improvements to suggest, but this does feel like a data structure that is starting to get a bit unwieldy and may eventually hurt readability. nothing that's actionable atm, just a note as i look through this!

Reply:
I agree, I am not entirely happy with this either.  I think there's a path forward to make it cleaner when we remove the backwards-compatible logic for the old-style consent preferences.

I did experiment with splitting this out but have left them together at the moment for two reasons: 1) The two rounds of filtering on privacy preferences for saas connectors.  You want to make sure the second round doesn't get dropped. 2) `should_opt_in_to_service` also answers this question for old-style `consent` preferences. It feels weird to make relevant privacy preferences an argument here as that's only applicable for new-style preferences:



saas connector excerpt
```python
   def run_consent_request(
        self,
        node: TraversalNode,
        policy: Policy,
        privacy_request: PrivacyRequest,
        identity_data: Dict[str, Any],
        session: Session,
    ) -> bool:
    ...
    relevant_privacy_preferences: List[PrivacyPreferenceHistory] = filter_privacy_preferences_for_propagation(self.configuration.system, privacy_request)
    should_opt_in: Optional[bool] = should_opt_in_to_service(
        privacy_request, relevant_privacy_preferences=relevant_privacy_preferences
    )

    if should_opt_in is None:
        logger.info(
            "Skipping consent requests on node {}: No actionable consent preferences to propagate",
            node.address.value,
        )
        raise SkippingConsentPropagation(
            f"Skipping consent propagation for node {node.address.value} - no actionable consent preferences to propagate"
        )

    # Filter "relevant" preferences further in case there are conflicting preferences
    filtered_preferences = List[PrivacyPreferenceHistory] = [
        pref
        for pref in relevant_privacy_preferences
        if pref.preference == (UserConsentPreference.opt_in if should_opt_in else UserConsentPreference.opt_out)
    ]
```